packages/ckeditor5-link/src/linkcommand.ts

```diff
@@ -227,14 +231,21 @@ export default class LinkCommand extends Command {
 				}
 
 				for ( const range of rangesToUpdate ) {
-					writer.setAttribute( 'linkHref', href, range );
+					let linkRange = range;
+
+					if ( selection.getAttribute( 'linkHref' ) === linkText ) {
```

Comment:
In the text extraction helper you're checking the text under the first position of the first range in the selection. 
Here, the code is iterating over multiple possible ranges that can accept the link. 
And on top of that, you're obtaining `linkHref`'s value from selection's attribute (which is the attribute value under the selection's first position only).

Check out this use-case:

```js
editor.setData( '<p><a href="foo">foo</a></p><p>bar</p><p>baz</p>' )

editor.model.change( writer => writer.setSelection( editor.model.document.getRoot(), 'in' ) );

editor.execute( 'link', 'foo' );

console.log( editor.getData() );
```

You'd expect the 2nd and 3rd paragraphs to keep their text after linking but their text gets replaced by "foo".